File: rust/glacier/src/util/timer.rs
//! Repeating, cancellable timers.
//!
//! Glacier's [`Timer`]s allows to periodically emit [`Signal`].
//!
//! When a `Timer` starts, it emit a [`Started`] signal containing a handle to this timer.
//! When it stops a [`Stopped`] signal is emitted.
//!
//! By default, [`Timer`]s repeat in a loop until stopped, emitting a [`Timeout`] every-time they
//! expires.

use std::{
    ops::DerefMut,
    sync::{Arc, Mutex, Weak},
    time::Duration,
};

use snowcap_api::signal::{HandlerPolicy, Signal, Signaler};
use tokio::sync::mpsc;

#[derive(Clone, Copy)]
pub(crate) enum StopToken {
    Stop,
    Restart(bool),
}

type OnTimeoutCallback = Box<dyn Fn(&mut TimerData) + Send>;

pub struct TimerData {
    pub interval: Duration,
    pub repeat: bool,
    signaler: Signaler,
}

pub struct Inner {
    data: TimerData,
    on_timeout: Option<OnTimeoutCallback>,
    stop_token: Option<mpsc::UnboundedSender<StopToken>>,
}

/// Signal sent when a [`Timer`] starts.
#[derive(Clone, Signal)]
pub struct Started(pub Timer);

/// Signal sent when a [`Timer`] stops.
#[derive(Clone, Signal)]
pub struct Stopped(pub Timer);

/// Signal sent when a [`Timer`] timeout.
#[derive(Clone, Signal)]
pub struct Timeout(pub Timer);

/// Repeating, cancellable timer.
#[derive(Clone)]
pub struct Timer {
    handle: Arc<Mutex<Inner>>,
}

/// Non-owning reference to a timer.
#[derive(Clone)]
pub struct WeakTimer(Weak<Mutex<Inner>>);

impl Timer {
    /// Create a new repeating timer from a given interval and an existing Signaler.
    pub fn new_with_signaler(interval: Duration, signaler: Signaler) -> Self {
        let data = TimerData {
            interval,
            repeat: true,
            signaler,
        };

        let handle = Arc::new(Mutex::new(Inner {
            data,
            on_timeout: None,
            stop_token: None,
        }));

        Self { handle }
    }

    /// Create a new repeating timer from a given interval.
    pub fn new(interval: Duration) -> Self {
        Self::new_with_signaler(interval, Signaler::new())
    }

    pub fn once_with_signaler(timeout: Duration, signaler: Signaler) -> Self {
        let mut ret = Self::new_with_signaler(timeout, signaler);
        ret.repeat(false);

        ret
    }

    /// Create a non-repeating timer.
    pub fn once(timeout: Duration) -> Self {
        Self::once_with_signaler(timeout, Signaler::new())
    }

    /// Sets the timer interval.
    pub fn interval(&mut self, interval: Duration) {
        self.handle.lock().unwrap().data.interval = interval;
    }

    /// Make this timer repeating or non-repeating.
    pub fn repeat(&mut self, repeat: bool) {
        self.handle.lock().unwrap().data.repeat = repeat;
    }

    /// Connect a [`Signal`] handler to this timer.
    pub fn connect<Sig, F>(&mut self, callback: F)
    where
        Sig: Signal + Clone,
        F: Fn(Sig) -> HandlerPolicy + Send + Sync + 'static,
    {
        self.handle.lock().unwrap().data.signaler.connect(callback);
    }

    /// Set a callback to be called on timeout.
    ///
    /// The callback will be called with the [`TimerData`], allowing it to manage the current state
    /// or emit custom signals in addition to the standard Timeout.
    pub fn on_timeout<F>(&mut self, callback: F)
    where
        F: Fn(&mut TimerData) + Send + 'static,
    {
        self.handle.lock().unwrap().on_timeout = Some(Box::new(callback));
    }

    /// Create a [`WeakTimer`] from this timer.
    pub fn downgrade(&self) -> WeakTimer {
        WeakTimer(Arc::downgrade(&self.handle))
    }

    /// Starts the timer.
    ///
    /// If `now` is set, a [`Timeout`] signal is sent right after the [`Started`] signal.
    ///
    /// Calling start on a running timer does nothing.
    pub fn start(&self, now: bool) {
        if self.handle.lock().unwrap().stop_token.is_some() {
            return;
        }

        let (send, mut recv) = mpsc::unbounded_channel();

        self.handle.lock().unwrap().stop_token = Some(send);

        let weak = self.downgrade();
        tokio::spawn(async move {
            {
                let Some(handle) = weak.upgrade() else {
                    return;
                };

                handle.with_signaler(|s| s.emit(Started(handle.clone())));

                if now {
                    handle.process_timeout();
                }
            }

            loop {
                let Some(handle) = weak.upgrade() else {
                    return;
                };

                let interval = handle.with_data(|d| d.interval);

                let (timeout, restart) = match tokio::time::timeout(interval, recv.recv()).await {
                    Ok(Some(StopToken::Restart(now))) => (now, true),
                    Ok(None) | Ok(Some(StopToken::Stop)) => {
                        handle.with_data_mut(|data| data.repeat = false);
                        (false, false)
                    }
                    Err(_) => (true, false),
                };

                if timeout {
                    handle.process_timeout();
                }

                if handle.with_data(|d| !d.repeat) {
                    handle.with_signaler(|s| s.emit(Stopped(handle.clone())));
                    break;
                } else if restart {
                    handle.with_signaler(|s| s.emit(Started(handle.clone())));
                }
            }
        });
    }

    /// Stops this timer.
    pub fn stop(&self) {
        let mut guard = self.handle.lock().unwrap();

        if let Some(stopper) = guard.stop_token.take() {
            let _ = stopper.send(StopToken::Stop);
        }
    }

    /// Restart this timer.
    ///
    /// This is similar to calling [`stop`] followed by [`start`], except it won't emit a
    /// [`Stopped`] signal if the timer was running.
    ///
    /// [`stop`]: Timer::stop
    /// [`start`]: Timer::start
    pub fn restart(&self, now: bool) {
        let guard = self.handle.lock().unwrap();

        if let Some(stopper) = &guard.stop_token {
            let _ = stopper.send(StopToken::Restart(now));
        } else {
            drop(guard);
            self.start(now);
        }
    }

    fn process_timeout(&self) {
        let mut guard = self.handle.lock().unwrap();
        let Inner {
            data, on_timeout, ..
        } = guard.deref_mut();

        data.signaler.emit(Timeout(self.clone()));

        if let Some(on_timeout) = on_timeout {
            on_timeout(data);
        }
    }

    fn with_signaler<F>(&self, processor: F)
    where
        F: FnOnce(&Signaler),
    {
        let guard = self.handle.lock().unwrap();

        processor(&guard.data.signaler)
    }

    fn with_data<F, Ret>(&self, processor: F) -> Ret
    where
        F: FnOnce(&TimerData) -> Ret,
    {
        let guard = self.handle.lock().unwrap();

        processor(&guard.data)
    }

    fn with_data_mut<F, Ret>(&self, processor: F) -> Ret
    where
        F: FnOnce(&mut TimerData) -> Ret,
    {
        let mut guard = self.handle.lock().unwrap();

        processor(&mut guard.data)
    }
}

impl WeakTimer {
    /// Attempts to upgrade the `WeakTimer` to a [`Timer`].
    ///
    /// Returns [`None`] if the timer was already dropped.
    pub fn upgrade(&self) -> Option<Timer> {
        self.0.upgrade().map(|handle| Timer { handle })
    }
}

impl TimerData {
    pub fn signaler(&self) -> &Signaler {
        &self.signaler
    }
}

File: lua/glacier/utils/timer.lua
local cprom = require("cqueues.promise") --[[@as cqueues.promiselib]]

local event_loop = require("glacier.internals.event_loop")

---glacier.timer module
---
---@class glacier.timer
---@field mt metatable This module metatable
---
---@overload fun(...: glacier.timer.Config):glacier.timer.Timer
local timer = { mt = {} }

---Signals sent by timers
---@enum glacier.timer.signals
local signals = {
    ---Emitted when a Timer start or re-start.
    STARTED = "glacier::timer::started",
    ---Emitted when a Timer stops.
    STOPPED = "glacier::timer::stopped",
    ---Emitted when a timeout occurs.
    TIMEOUT = "glacier::timer::timeout",
}

---Cancellable sleep function.
---
---@param timeout number
---@param promise cqueues.promise
---@return boolean # Return true if the timeout woke us up, false if the function was cancelled.
---@return cqueues.promise? # If set, the timer should be immediately restarted using this promise instead
local function promise_sleep(timeout, promise)
    local ret = promise:wait(timeout)

    if ret then
        return false, promise:get(0)
    end

    return true, nil
end

---@class glacier.timer.Config
---
---The signaler to use. If left empty, a new Signaler will be created
---@field signaler? snowcap.signal.Signaler
---The repetition interval for this signaler. Defaults to 1.0.
---@field interval? number
---Whether this timer should repeat. Defaults to true.
---@field repeats? boolean
---Callback to be called on timeout.
---
---The callback will be called along with the [`TimerData`], allowing it to manage the current state
---or emit custom signals in addition to the standard Timeout.
---@field on_timeout? fun(data: glacier.timer.TimerData)

---Internal state of a timer.
---
---@class glacier.timer.TimerData
---@field interval number Amount of time between events.
---@field repeats boolean If true, this timer stop itself after timing out.
---@field signaler snowcap.signal.Signaler

---A simple timer.
---
---`Timer`s can be used to defer some actions, or run some callback periodically.
---
---When the internal timeout expires, a `Timer` will emit
---`glacier.timer.signals.TIMEOUT` and call their on_timeout callback if set.
---If the `Timer` was created without the `repeats` flag set, or via
---`glacier.timer.once()` it will stop after sending the event. Otherwise, the
---`Timer` will wait again in a loop until stopped.
---
---When a timer stop, `glacier.timer.signals.STOPPED` is emitted.
---@class glacier.timer.Timer
---@field private data glacier.timer.TimerData
---@field private on_timeout fun(data: glacier.timer.TimerData)|nil If present, a function to call when the timer timeouts.
---@field private promise cqueues.promise|nil Promise to control the timer.
local Timer = {}

----------------------------
-- Timer's public methods --
----------------------------

---Starts the Timer.
---
---Once started, the timer will periodically emit `timer::timeout`.
---
---@param now? boolean If true, fire an event before sleep.
function Timer:start(now)
    assert(self.promise == nil, "Timer::start() should not be called on a running timer.")

    self.promise = cprom.new()
    event_loop.loop:wrap(function()
        self.data.signaler:emit(signals.STARTED, self)

        if now then
            self:process_timeout()
        end

        local continue = true
        while continue do
            local timeout, restart = promise_sleep(self.data.interval, self.promise)

            if timeout then
                self:process_timeout()
            elseif restart then
                self.promise = restart

                if self.data.repeats then
                    self:emit(signals.STARTED)
                end
            else
                continue = false
            end

            continue = continue and self.data.repeats
        end

        self:emit(signals.STOPPED)
    end)
end

---Stops the Timer.
function Timer:stop()
    if self.promise then
        self.promise:set(true, nil)
        self.promise = nil
    end
end

---Restarts the timer.
---
---This is equivalent of calling stop if the timer was started, followed by a call to start.
---
---@param now? boolean If true, fire an event synchronously.
function Timer:restart(now)
    if self.promise then
        if now then
            self:emit(signals.TIMEOUT)
        end

        self.promise:set(true, cprom.new())
    else
        self:start(now)
    end
end

---Connect to the timer signals.
---
---This is a shorthand for `Timer:signaler():connect()`.
---
---@param sig string
---@param callback any Callback to call when the signal fire.
---@return snowcap.signal.SignalHandle # A handle to disconnect this callback.
function Timer:connect(sig, callback)
    return self.data.signaler:connect(sig, callback)
end

---Access the timer `Signaler`.
---
---@return snowcap.signal.Signaler
function Timer:signaler()
    return self.data.signaler()
end

-----------------------------
-- Timer's private methods --
-----------------------------

---@private
---
---Emit standard signals.
---
---@param sig string
function Timer:emit(sig)
    self.data.signaler:emit(sig, self)
end

---@private
---
---Function called on timeout.
function Timer:process_timeout()
    self:emit(signals.TIMEOUT)

    if self.on_timeout then
        self.on_timeout(self.data)
    end
end

-----------
-- Other --
-----------

---Create a new timer.
---
---@param config glacier.timer.Config
---@return glacier.timer.Timer
function Timer:new(config)
    config = config or {}
    config.repeats = config.repeats == nil and true or config.repeats

    ---@type glacier.timer.Timer
    local ret = {
        data = {
            interval = config.interval or 1.0,
            repeats = config.repeats,
            signaler = config.signaler or require("snowcap.signal").Signaler.new(),
        },
        on_timeout = config.on_timeout,
        promise = nil,
    }
    ret = setmetatable(ret, { __index = Timer })

    return ret
end

---Create a new timer
---
---@param ... glacier.timer.Config
---@return glacier.timer.Timer
function timer.mt:__call(...)
    return Timer:new(...)
end

---Create a timer that will fire only once.
---
---The timer is tstarted before being returned by this function.
---
---@param callback fun(data: glacier.timer.TimerData) Function to call when the timer expires.
---@param timeout number How long to wait before running the callback.
---
---@return glacier.timer.Timer
function timer.once(callback, timeout)
    local ret = Timer:new({
        interval = timeout or 0.0,
        on_timeout = callback,
        repeats = false,
    })

    ret:start()

    return ret
end

---Create a new timer and starts it immediately.
---
---@param ... glacier.timer.Config Configuration options.
---@return glacier.timer.Timer
function timer.started(...)
    local ret = Timer:new(...)

    ret:start()

    return ret
end

timer.Timer = Timer
timer.signals = signals

---@diagnostic disable-next-line: param-type-mismatch
return setmetatable(timer, timer.mt) --[[@as glacier.timer]]
